packages/convex-helpers/server/zod4.ts

```diff
@@ -0,0 +1,642 @@
+import { z } from "zod";
+import type {
+  GenericId,
+  PropertyValidators,
+  GenericValidator,
+  Value,
+} from "convex/values";
+import { ConvexError, v } from "convex/values";
+import type {
+  FunctionVisibility,
+  GenericDataModel,
+  GenericActionCtx,
+  GenericQueryCtx,
+  MutationBuilder,
+  QueryBuilder,
+  GenericMutationCtx,
+  ActionBuilder,
+  TableNamesInDataModel,
+  DefaultFunctionArgs,
+} from "convex/server";
+import type { Customization } from "./customFunctions.js";
+import { NoOp } from "./customFunctions.js";
+import { pick } from "../index.js";
+import { addFieldsToValidator } from "../validators.js";
+
+export type ZodValidator = Record<string, z.ZodTypeAny>;
+
+// Simple registry for zid metadata
+const _meta = new WeakMap<z.ZodTypeAny, { isConvexId?: boolean; tableName?: string }>();
+const registryHelpers = {
+  getMetadata: (schema: z.ZodTypeAny) => _meta.get(schema),
+  setMetadata: (schema: z.ZodTypeAny, meta: { isConvexId?: boolean; tableName?: string }) =>
+    _meta.set(schema, meta),
+};
+
+export function zid<
+  DataModel extends GenericDataModel,
+  TableName extends TableNamesInDataModel<DataModel> = TableNamesInDataModel<DataModel>,
+>(tableName: TableName) {
+  const base = z
+    .string()
+    .refine((s) => typeof s === "string" && s.length > 0, {
+      message: `Invalid ID for table "${tableName}"`,
+    })
+    .transform((s) => s as GenericId<TableName>)
+    .brand(`ConvexId_${tableName}`)
+    .describe(`convexId:${tableName}`);
+  registryHelpers.setMetadata(base, { isConvexId: true, tableName });
+  return base as z.ZodType<GenericId<TableName>>;
+}
+
+export const withSystemFields = <
+  Table extends string,
+  T extends { [key: string]: z.ZodTypeAny },
+>(tableName: Table, zObject: T) => {
+  return { ...zObject, _id: zid(tableName as any), _creationTime: z.number() } as const;
```

Comment:
I think that `as any` here can be avoided
```suggestion
  return { ...zObject, _id: zid(tableName), _creationTime: z.number() } as const;
```